### DISCUSSION TITLE
Is it possible to use react node for decoration?

### DISCUSSION THREAD

**Discussion by pie6k** [ueberdosis/tiptap#2285] [General Questions]
I'd like to use something very similar to placeholder, but showing "+" icon next to the cursor in empty nodes:

<img width="358" alt="CleanShot 2021-12-16 at 14 05 12@2x" src="https://user-images.githubusercontent.com/7311462/146377241-35622770-41ea-47da-bdf9-8d50514dbd30.png">

I imagine this could be achieved without React with bunch of CSS hacking or vanilla JS DOM manipulations, but in React I have both icon and tooltip component in place.

Thus question if it is possible to create react-rendered nodes for decorations?

**Comment by omarkhairy21** [comment]
i have done something similar using floating-menu https://tiptap.dev/api/extensions/floating-menu and yon can position it from tippyjs 

```
<FloatingMenu editor={editor} tippyOptions={{ zIndex: 1, placement: 'left' }}> 
...
 </FloatingMenu>

```

**Comment by Aldredcz** [comment]
Any plans on adding support for using React code for decorations? It sounds to me like a very important part for building versatile editors. Or any other approaches are recommended?

I see the approach described by you, @omarkhairy21, but I think it comes short when you need to render a decoration (widget / node) for every block, not just active/selected one.

Thank you! 🙏

**Comment by fdintino** [comment]
This question is nearly a year old, but in case it is at all helpful: it's possible to render react elements in a decoration with react-dom. You'd need to use `Decoration.widget` so you have a place to call the unmount method.

```tsx
import ReactDOM from "react-dom/client";
import { Decoration } from "@tiptap/pm/view";

let root: ReactDOM.Root | null = null;

const decoration = Decoration.widget(
  pos,
  () => {
    const div = document.createElement("div");
    root = ReactDOM.createRoot(div);
    root.render(<MyComponent />);
    return div;
  },
  {
    destroy: () => root?.unmount(),
  },
);

**Comment by mweidner037** [comment]
There is an example using React Portals to render a React component inside a decoration, in the new AI toolkit docs: https://tiptap.dev/docs/content-ai/capabilities/ai-toolkit/advanced-guides/style-suggestions#render-react-components-inside-a-suggestion

To get this working outside of that specific example (AI suggestions), you'd need a react component above the editor that:
- Stores & renders the portal(s)
- Passes a setPortalElement (or add/deletePortalElement) function into the extension that creates the decorations.
